two-scale-heat-conduction/macro-nutils/macro.py

```diff
@@ -0,0 +1,183 @@
+#! /usr/bin/env python3
+#
+# Heat conduction problem with two materials.
+
+from nutils import mesh, function, solver, export, cli
+import treelog
+import numpy as np
+import precice
+
+
+def main():
+    """
+    2D heat conduction problem solved on a rectangular domain. The material consists of a mixture of two materials, the grain (g) and sand (s).
+    """
+    is_coupled_case = True  # Change to False to solve single physics case without macro-micro coupling
+
+    # Original case from Bastidas et al.
+    # topo, geom = mesh.rectilinear([np.linspace(0, 1.0, 9), np.linspace(0, 0.5, 5)])
+
+    topo, geom = mesh.rectilinear([np.linspace(0, 1.0, 4), np.linspace(0, 0.5, 3)])
+
+    ns = function.Namespace(fallback_length=2)
+    ns.x = geom
+    ns.basis = topo.basis('std', degree=1)
+    ns.u = 'basis_n ?solu_n'  # the field to be solved for
+
+    ns.phi = 'basis_n ?solphi_n'  # porosity, to be read from preCICE
+    phi = 0.5  # initial value of porosity
+
+    ns.kbasis = topo.basis('std', degree=1).vector(topo.ndims).vector(topo.ndims)
+    # conductivity matrix, to be read from preCICE (read as two vectors and then patched together)
+    ns.k_ij = 'kbasis_nij ?solk_n'
+    k = 1.0  # initial value of conductivity
+
+    ns.rhos = 1.0  # density of material s
+    ns.rhog = 1.0  # density of material g
+    ns.dudt = 'basis_n (?solu_n - ?solu0_n) / ?dt'  # implicit Euler time stepping formulation
+
+    # Dirichlet boundary condition on lower left corner
+    ns.usource = 0.0
+
+    # Initial condition in the domain, except the lower left corner
+    ns.uinitial = 0.5
+
+    if is_coupled_case:
+        interface = precice.Interface("Macro-heat", "precice-config.xml", 0, 1)
```

Comment:
The `README.md` tells me to just run `./run.sh`. However, if I do that, it complains that:

```
(0) 11:08:41 [xml::XMLParser]:152 in readXmlFile: ERROR: XML parser was unable to open configuration file "precice-config.xml"
```
I guess it should be:

```suggestion
        interface = precice.Interface("Macro-heat", "../precice-config.xml", 0, 1)
```